modules/vorbis/audio_stream_ogg_vorbis.h

```diff
@@ -0,0 +1,213 @@
+
+	friend class OggPacketSequencePlayback;
+
+	// A map from page index to a vector of the raw packet data for each packet in that page.
+	Map<int64_t, Vector<Vector<uint8_t>>> page_data;
+
+	// A map from page index to a vector of the packet metadata for each packet in that page (in order: b_o_s, e_o_s, granulepos, packetno).
+	Map<int64_t, Vector<Vector<int64_t>>> page_metadata;
+
+	// The page after the current last page. Similar semantics to an end() iterator.
+	int64_t end_page = 0;
+
+	uint64_t data_version = 0;
+
+	float sampling_rate = 0;
+	float length = 0;
+
+protected:
+	static void _bind_methods();
+
+public:
+	// Pushes information about all the pages that ended on this page.
+	// This should be called for each page, even for pages that no packets ended on.
+	void push_page(const Vector<PackedInt64Array> &p_metadata, const Vector<PackedByteArray> &p_data);
+
+	void set_packet_data(const Dictionary &p_data);
+	Dictionary get_packet_data() const;
+
+	void set_packet_metadata(const Dictionary &p_metadata);
+	Dictionary get_packet_metadata() const;
+
+	// Sets a sampling rate associated with this object. OggPacketSequence doesn't understand codecs,
+	// so this value is naively stored as a convenience.
+	void set_sampling_rate(float p_sampling_rate);
+
+	// Returns a sampling rate previously set by set_sampling_rate().
+	float get_sampling_rate() const;
+
+	// Sets a length in seconds associated with this object. OggPacketSequence doesn't understand codecs,
+	// so this value is naively stored as a convenience.
+	void set_length(float p_length);
+
+	// Returns a length previously set by set_length().
+	float get_length() const;
+
+	// Returns the granule position of the last page in this sequence.
+	int64_t get_final_granule_pos() const;
+
+	Ref<OggPacketSequencePlayback> instance_playback();
+
+	OggPacketSequence() {}
+	virtual ~OggPacketSequence() {}
+};
+
+class OggPacketSequencePlayback : public RefCounted {
+	GDCLASS(OggPacketSequencePlayback, RefCounted);
+
+	friend class OggPacketSequence;
+
+	Ref<OggPacketSequence> ogg_packet_sequence;
+
+	mutable int64_t page_cursor = 0;
+	mutable int32_t packet_cursor = 0;
+
+	mutable ogg_packet packet;
+
+	uint64_t data_version;
+
+	// Recursive bisection search for the correct page.
+	uint32_t seek_page_internal(int64_t granule, uint32_t after_page_inclusive, uint32_t before_page_inclusive);
+
+public:
+	// Calling functions must not modify this packet.
+	// Returns true on success, false on error or if there is no next packet.
+	bool next_ogg_packet(ogg_packet **p_packet) const;
+
+	// Seeks to the page such that the previous page has a granule position less than or equal to this value,
+	// and the current page has a granule position greater than this value.
+	// Returns true on success, false on failure.
+	bool seek_page(int64_t p_granule_pos);
+
+	OggPacketSequencePlayback() {}
+	virtual ~OggPacketSequencePlayback() {}
+};
+
+class AudioStreamOggVorbis;
```

Comment:
The current classes have `OGG`. I'm not sure which is desired (is it even an acronym...?), but this needs discussion.